2_process/src/add_dynamic_attr.R

```diff
@@ -0,0 +1,796 @@
+}
+
+
+compute_lagged_attrs <- function(dyn_df, lag_table, lag_attr_name, 
+                                 attr_name, start_date){
+  #' @description computes lagged predictor variables from the provided features
+  #' 
+  #' @param dyn_df tbl containing "Date" and the PRMS "seg". Should have all
+  #' dates that you want to use for lagged computations.
+  #' @param lag_table table with 3 columns: attributes, lags, and lag_unit.
+  #' This is used to compute the lagged attributes
+  #' @param lag_attr_name name in the attribute column of lag_table to use
+  #' @param attr_name name to use for this attribute. Format of the new name will
+  #' be paste0(attr_prefix, '_', attr_name, '_', lags, lag_unit)
+  #' @param start_date first date to compute attributes
+  #' 
+  #' @return tbl with the added lagged features. Date begins at the start_date
+  #' and ends at the latest provided date in dyn_df.
+  
+  #Get the lag information and lag functions to be used
+  tmp_lags <- lag_table %>% 
+    filter(attribute == lag_attr_name) %>%
+    pull(lags) %>%
+    unlist()
+  tmp_lag_unit <- lag_table %>% 
+    filter(attribute == lag_attr_name) %>%
+    pull(lag_unit) %>%
+    unlist()
+  #Make the length of the lag_unit match the length of lags
+  if ((length(tmp_lag_unit) == 1) & (length(tmp_lags) > 1)){
+    tmp_lag_unit <- rep(tmp_lag_unit, length(tmp_lags))
+  }
+  tmp_lag_fxns <- lag_table %>% 
+    filter(attribute == lag_attr_name) %>%
+    pull(lag_fxns) %>%
+    unlist()
+  
+  if (length(tmp_lag_fxns) > 0){
+    #Loop over all functions except exact
+    for (f in 1:length(tmp_lag_fxns)){
+      if (tmp_lag_fxns[f] != 'exact'){
+        #Loop over all lags
+        for (lag in 1:length(tmp_lags)){
+          #get the number of days to lag
+          days_lag <- as.Date(start_date) - 
+            (as.Date(start_date) - period(tmp_lags[lag], tmp_lag_unit[lag]))
+          
+          #Compute the lagged values
+          if (tmp_lag_fxns[f] == 'mean'){
+            dyn_df <- cbind(dyn_df, 
+                            dyn_df %>% 
+                              group_by(seg) %>% 
+                              summarise(across(.cols = ends_with(paste0(attr_name, '_0')), 
+                                               .fns = stats::filter, 
+                                               filter = rep(1/as.numeric(days_lag), days_lag), 
+                                               sides = 1, 
+                                               .names = paste0("{.col}_", tmp_lags[lag], 
+                                                               tmp_lag_unit[lag], "_mean")), 
+                                        .groups = 'drop') %>%
+                              select(-seg) %>%
+                              #need to shift the results by 1 day because filter 
+                              #includes the current day in the lag
+                              add_row(.before = 1) %>%
+                              slice(-nrow(.))
```

Comment:
Thanks! I appreciate all of the optimization suggestions. I might try the foreach loop because I'll have to rebuild these targets after making the PR changes, anyway. And I know that those loops are the main time consuming steps